relay-kafka/src/producer.rs

```diff
@@ -0,0 +1,170 @@
+use std::{collections::BTreeMap, sync::Arc};
+
+use rdkafka::{producer::BaseRecord, ClientConfig};
+
+use relay_statsd::metric;
+
+use crate::{
+    config::{KafkaConfig, KafkaParams},
+    utils::{CaptureErrorContext, KafkaHistograms, ThreadedProducer},
+    Message, ProducerError,
+};
+
+/// Temporary map used to deduplicate kafka producers
+type ReusedProducersMap<'a> = BTreeMap<Option<&'a str>, Arc<ThreadedProducer>>;
+
+/// This object containes the Kafka producer variants for single and sharded configurations.
+pub enum Producer {
```

Comment:
_nit_: The internals of the enum are now part of the public signature. Do we want this explicitly?

- If yes, then let's also introduce a type for the `Single` variant. This allows us to move specific functionality to a method on that type rather than having large functions covering all variants here.
- If not, let's make this a `struct` with a hidden `inner` field that is then a private enum. I'd still propose the refactor from the previous point, though.